rfcs/20200113-tf-data-service.md

```diff
@@ -0,0 +1,591 @@
+  else:
+    iteration_id = broadcast_recv()
+  it = tf.data.experimental.service.make_iterator(
+      ds, iteration_id, consumer_index)
+  for element in it:
+    # process element
+
+  Args:
+    dataset: The dataset to begin iteration over.
+    num_consumers: The number of consumers to divide the dataset between. Set
+      this if you require determinism. If None, a single iterator id is returned,
+      and any number of consumers can read from that iterator id. The data
+      produced by the dataset will be fed to consumers on a first-come
+      first-served basis.
+    num_tasks: The number of tasks to use for processing. Tasks run for
+      the duration of an epoch, and each worker should typically process a single
+      task. Normally it is best to leave this as None so that the master can
+      choose a reasonable number of tasks. Setting `num_tasks` is useful for
+      producing deterministic results.
+    deterministic: Whether the iteration should be performed
+      deterministically. Fully deterministic output also requires setting
+      `num_tasks` to a fixed number, and that the input dataset is itself
+      deterministic.
+
+  Returns:
+    An iteration_id which can be used to created iterators via
+      `tf.data.experimental.service.make_iterator`
+  """
+
+def tf.data.experimental.service.make_iterator(
+    dataset, iteration, consumer_index):
+  """Creates an iterator for reading from the specified dataset.
+
+  Args:
+    dataset: The dataset to read from.
+    iteration: An iteration_id object generated by
+      `tf.data.experimental.service.create_iteration`.
+    consumer_index: The consumer index within the iteration to read from. If
+      the iteration was created with `n` consumers, `consumers_index` must be
+      less than `n`.
+
+  Returns:
+    A Python iterator which iterates over the dataset elements.
+  """
+```
+
+### Dataset splitting API
+
+To parallelize dataset processing, the tf.data service needs a way to split up
+datasets. We will achieve this by adding a splitting API that allows source
+datasets to express how they can be split.
+
+Our goals for the API are
+
+*   Performance: The splitting API can be used to performantly split and process
+    datasets.
+*   Extensibility: User-defined datasets can be split as long as they implement
+    the splitting API.
+*   Minimize Surprises: Users write their datasets as though they will not be
+    split, so introducing splitting can easily lead to unexpected outcomes. To
+    mitigate this, we will be conservative about which dataset transformations
+    support splitting.
+
+The API will be used internally by the tf.data service to distribute datasets.
+It will be entirely in C++, and we don't currently have any plans to expose
+splitting through Python.
+
+The API focuses on producing and consuming `Split`s. A `Split` is a variant
+Tensor that can be subclassed to represent arbitrary types of splitting.
+
+```cpp
+class Split {
```

Comment:
This is very vague. What is stored in the variant?

Reply:
The interface is intentionally vague so that implementors can define their splits however makes the most sense for their dataset type.

Reply:
Added this comment to the doc